pinot-core/src/main/java/org/apache/pinot/core/query/reduce/BrokerReduceService.java

```diff
@@ -392,6 +434,218 @@ private void setAggregationResults(@Nonnull BrokerResponseNative brokerResponseN
+  }
+
+  private IndexedTable getIndexedTable(int numGroupBy, int numAggregations, GroupBy groupBy,
+      List<AggregationInfo> aggregationInfos, List<SelectionSort> orderBy, DataSchema dataSchema, Map<ServerInstance, DataTable> dataTableMap)
+      throws Throwable {
+
+    IndexedTable indexedTable = new ConcurrentIndexedTable();
+    // setting a higher value to avoid frequent resizing
+    int capacity = (int) Math.max(groupBy.getTopN(), 10000);
+    indexedTable.init(dataSchema, aggregationInfos, orderBy, capacity, true);
+
+    for (DataTable dataTable : dataTableMap.values()) {
+      CheckedFunction2[] functions = new CheckedFunction2[dataSchema.size()];
+      for (int i = 0; i < dataSchema.size(); i++) {
+        ColumnDataType columnDataType = dataSchema.getColumnDataType(i);
+        CheckedFunction2<Integer, Integer, Object> function;
+        switch (columnDataType) {
+
+          case INT:
+            function = (CheckedFunction2<Integer, Integer, Object>) dataTable::getInt;
+            break;
+          case LONG:
+            function = (CheckedFunction2<Integer, Integer, Object>) dataTable::getLong;
+            break;
+          case FLOAT:
+            function = (CheckedFunction2<Integer, Integer, Object>) dataTable::getFloat;
+            break;
+          case DOUBLE:
+            function = (CheckedFunction2<Integer, Integer, Object>) dataTable::getDouble;
+            break;
+          case STRING:
+            function = (CheckedFunction2<Integer, Integer, Object>) dataTable::getString;
+            break;
+          default:
+            function = (CheckedFunction2<Integer, Integer, Object>) dataTable::getObject;
+        }
+        functions[i] = function;
+      }
+
+      for (int row = 0; row < dataTable.getNumberOfRows(); row++) {
+        Object[] key = new Object[numGroupBy];
+        int col = 0;
+        for (int j = 0; j < numGroupBy; j++) {
+          key[j] = functions[col].apply(row, col);
+          col ++;
+        }
+        Object[] value = new Object[numAggregations];
+        for (int j = 0; j < numAggregations; j++) {
+          value[j] = functions[col].apply(row, col);
+          col ++;
+        }
+        Record record = new Record(new Key(key), value);
+        indexedTable.upsert(record);
+      }
+    }
+    indexedTable.finish();
+    return indexedTable;
+  }
+
+  /**
+   * Extract the results of group by order by into a List of {@link AggregationResult}
+   * There will be 1 aggregation result per aggregation. The group by keys will be the same across all aggregations
+   * @param brokerResponseNative broker response
+   * @param dataSchema data schema
+   * @param aggregationInfos aggregations info
+   * @param groupBy group by info
+   * @param orderBy order by info
+   * @param dataTableMap map from server to data table
+   */
+  private void setPQLGroupByOrderByResults(@Nonnull BrokerResponseNative brokerResponseNative,
```

Comment:
Is this needed because we want to support order-by with PQL?

Reply:
this is required for responseFormat=pql